language/stdlib/modules/libra_account.mvir

```diff
@@ -57,15 +58,99 @@ module LibraAccount {
         payer: address,
     }
 
+    /// Events
+    // A resource representing the counter used to generate uniqueness under each account. There won't be destructor for
+    // this resource to guarantee the uniqueness of the generated handle.
+    resource EventHandleGenerator {
+        // A monotonically increasing counter
+        counter: u64,
+    }
+
+    // A handle for a event such that:
+    // 1. Other modules can emit event to this handle.
+    // 2. Storage can use this handle to prove the total number of events that happened in the past.
+    resource EventHandle<T: unrestricted> {
+        // Total number of events emitted to this event stream.
+        counter: u64
+        // A globally unique ID for this event stream.
+        guid: bytearray,
+    }
+
+    // Derive a fresh unique id by using sender's EventHandleGenerator. The generated bytearray is indeed unique because it
+    // was derived from the hash(sender's EventHandleGenerator || sender_address). This module guarantees that the
+    // EventHandleGenerator is only going to be monotonically increased and there's no way to revert it or destroy it. Thus
+    // such counter is going to give distinct value for each of the new event stream under each sender. And since we
+    // hash it with the sender's address, the result is guaranteed to be globally unique.
+    fresh_guid(counter: &mut Self.EventHandleGenerator): bytearray {
+        let sender: address;
+        let count: &mut u64;
+        let sender_bytes: bytearray;
+        let count_bytes: bytearray;
+        let preimage: bytearray;
+
+        sender = get_txn_sender();
+        count = &mut move(counter).counter;
+
+        sender_bytes = AddressUtil.address_to_bytes(move(sender));
+
+        count_bytes = U64Util.u64_to_bytes(*copy(count));
+        *move(count) = *copy(count) + 1;
+
+        // EventHandleGenerator goes first just in case we want to extend address in the future.
+        preimage = BytearrayUtil.bytearray_concat(move(count_bytes), move(sender_bytes));
```

Comment:
Just to be clear, the only change here is not doing a hash?

Also, I would have guessed it would easier to put the sender bytes first?

Reply:
Ahh I forgot to change the sender logic. My bad. For the sender bytes, I put the salt first for a specific reason. The main concern is if we are going to extend our address space to something bigger. If sth like that happens, having the sender in the front will make the upgrade a lot harder.